docs/naps/6-contributable-menus.md

```diff
@@ -0,0 +1,517 @@
+      python_name: napari_demo:submenu_item
+
+  menus:
+    napari/layers/context:
+      - submenu: context_submenu
+      - command: napari-demo.submenu_item
+    hello_world:
+      - command: napari-demo.menu_item
+
+  submenus:
+    - id: context_submenu
+      label: A new submenu
+```
+
+This NAP proposes new menu IDs and new top level menus to open for contribution.
+
+![Right click layer context menu with new menu item and submenu contributed by a plugin](./_static/layer-context-menu.png)
+
+
+### What do Menu Contributions do?
+
+`MenuItem` contributions can be thought of as auxiliary contributions that
+provide a dispatch mechanism for other existing contributions. Currently
+these would strictly be `widget` contributions, but this mechanism
+can easily be extended to other commands, which can take as input 
+`napari` objects like specific layers, or the `Viewer`, and produce
+output the `Viewer` uses - currently this would be new layers.
+
+Moving forward, new contribution types could be defined that allow
+plugin developers to run context aware commands that interact with 
+different `Viewer` components without the need for a `widget`.
+
+For example, `LayerEditor` contributions could take the currently
+active/selected `Layers` and edit the underlying data, while `LayerGenerator`
+contributions could take the same input and create new layers in
+the viewer. By providing dedicated contributions for such actions,
+`napari` can enforce rules about layer editing and layer 
+generation more strictly than for widget contributions,
+which, if taking the `Viewer`, can perform arbitrary actions upon
+all objects within the `Viewer`.
+
+We therefore propose a menu structure that would be easily
+extensible with these new contribution types and provide intuitive
+locations for both plugin developers to add their functionality,
+and users to find it. 
+
+## Detailed Description
+
+<!-- This section should provide a detailed description of the proposed change. It
+should include examples of how the new functionality would be used, intended
+use-cases, and pseudocode illustrating its use. -->
+
+We propose an initial set of contributable menus organized by the napari object
+being acted upon by the actions within the menu, and the likely output of those 
+actions. 
+
+### The `Layers` Menu
+Currently the foremost example of such an object is the napari `Layers`, and this
+menu therefore contains five submenus organized by the types of processing
+the user may wish to perform on the selected `Layer` or `Layers`.
+
+The `Layers` submenus are organized to give the user an immediate
+feeling of what might happen to their `Layers` as a result of clicking
+one of these menu items.
+
+1. `Visualization` - Items in this submenu allow you to generate visualizations from selected layer or layers.
+They do not change the layer data.
+2. `Measure` - Items in this submenu provide utilities for summarising information about your layer's data.
```

Comment:
What are the expected inputs/outputs for `Visualisation` and `Measure`?

Reply:
I think these menus are the hardest to categorize based on input/output. 

For `Visualization` I can imagine a widget that adds a new layer that provides a different "view" on the data. Mostly I was thinking of stuff like filtering what data is being shown in the current layer, auto adjustment of brightness/contrast, tools for quickly assigning/swapping preset colormaps... In general I think of it as "changing the way the layer is displayed in the viewer without changing the data of the layer".

`Measure` I think would typically contain items that generate text/summary output based on the layer, generate a new layer with information about the data (e.g. @psobolewskiPhD was mentioning he draws lines on `Shapes` layers to measure distances in `Image` layers or something like that - this would go here) or produce plots/figures about the layer data e.g. histograms, intensity plots, derived features, etc.

Reply:
> For Visualization I can imagine a widget that adds a new layer that provides a different "view" on the data. Mostly I was thinking of stuff like filtering what data is being shown in the current layer, auto adjustment of brightness/contrast, tools for quickly assigning/swapping preset colormaps... In general I think of it as "changing the way the layer is displayed in the viewer without changing the data of the layer".

This feels like a combination of `Generate` and `Edit` to me, though. 

> Measure I think would typically contain items that generate text/summary output based on the layer, generate a new layer with information about the data (e.g. @psobolewskiPhD was mentioning he draws lines on Shapes layers to measure distances in Image layers or something like that - this would go here) or produce plots/figures about the layer data e.g. histograms, intensity plots, derived features, etc.

Makes sense!

Reply:
> This feels like a combination of Generate and Edit to me, though.

Perhaps `Generate` if we include adding new layers (which is debatable tbh I don't actually know of any widgets that do this I just could imagine they do) but *definitely* not `Edit` - nothing in the `Visualization` submenu should *edit* the data of your layer, it should always be non-destructive. It should only change how the layer is displayed. For example a widget that filters tracks displayed in a tracks layer based on certain properties shouldn't be *deleting* tracks from the data - just turning their visibility on and off.

Reply:
But due to how visualisation and data in napari are still coupled in the `Layer` object, to change visualisation you *have* to edit the layer object. But I see what you mean, `Edit` should only affect the data (and maybe features?), and `Visualisation` only the rest.

With this in mind,`Visualisation` says that it "generates" visualisations, but it sounds like you mean "change visualisation parameters"? But then what about something that creates a plot widget like `napari-properties-plotter`? Shoudl that live in `Measure`?

Reply:
> But I see what you mean, Edit should only affect the data (and maybe features?), and Visualisation only the rest.

Yes, I do think there's a difference between the `Layer` data & features (which I think are intrinsically coupled as fundamental attributes) and other `Layer` attributes that determine visualization and other `napari` specific features. It's probably worth us defining a clear line for this as we look towards user facing documentation. 

> With this in mind, Visualisation says that it "generates" visualisations, but it sounds like you mean "change visualisation parameters"

Yes I'd say the second is closer to a correct description of what I have in mind. I will update this. Thank you!

> But then what about something that creates a plot widget like napari-properties-plotter? Shoudl that live in Measure?

I would say yes, as this is an auxiliary visualization, not changing the view of the layer itself.